--- base.py	2022-02-05 11:37:18.000000000 -0500
+++ base_cc.py	2022-02-05 11:35:10.000000000 -0500
@@ -234,10 +234,7 @@
 
         app = web.Application()
         self.sio.attach(app)
-        web.run_app(app,
-                    host=Config().server.address,
-                    port=port,
-                    loop=asyncio.get_event_loop())
+        web.run_app(app, host=Config().server.address, port=port)
 
     async def register_client(self, sid, client_id):
         """ Adding a newly arrived client to the list of clients. """
